### ISSUE TITLE
Include multiple API versions of an API group to a single clientset

### ISSUE THREAD

**Issue by caesarxuchao** [kubernetes/kubernetes#35237]
Forked from #35159

# Problem
Currently a generated golang [clientset](https://github.com/kubernetes/kubernetes/tree/master/pkg/client/clientset_generated/release_1_5) only includes code for one API version of each API group (e.g., [release_1_5 clientset](https://github.com/kubernetes/kubernetes/tree/master/pkg/client/clientset_generated/release_1_5) only supports [batch/v1](https://github.com/kubernetes/kubernetes/tree/master/pkg/client/clientset_generated/release_1_5/typed/batch/v1), but not batch/v2alpha1).

We designed it this way to spare users the trouble of selecting a version. However, this design is preventing us from migrating kubernetes components to use versioned clientset, for these reasons:

* It forces all components to move to use a new API version at the same time, defeating a main benefit of migrating to versioned objects.
* kube-controller-manager needs multiple versions. For example, the ScheduelJob controller needs to use v2alpha1.Job, but the Job controller needs to use v1.Job, otherwise `Job` stops working when admin turns off batch/v2alpha1 (but leaves batch/v1 on) on the api server.

# Proposed changes to clientset

A clientset will include multiple API versions of an API group. Its public interface will be changed:
* We'll add methods to select API version if a group have multiple versions, e.g., `clientset.BatchV1().Job()` returns `v1.Job`, and `clientset.BatchV2alpha1().Job()` returns `v2alpha1.Job`.
* If the caller doesn't pick a version, it will get the default version, e.g., `clientset.Batch().Job()` returns `v1.Job`, as it does today.
* If a group only has one version, we don't change its interface.

Regarding how many versions to include in a release of clientset, I currently plan to keep at most two stable versions of a group, and remove alpha or beta versions when they are removed from the kubernetes repository.

@kubernetes/sig-api-machinery @mml @liggitt

**Comment by caesarxuchao** [comment]
cc @nikhiljindal @smarterclayton @krousey

**Comment by caesarxuchao** [comment]
Including multiple versions in clientset will let the schedule job controller use clientset without hacking the client config: https://github.com/kubernetes/kubernetes/commit/7a34347f7f7503dea787c6cb0f27c5e97b1c49d2#diff-2aa9cfe6a01016c1679752b0fcfc9244R407. cc @soltysh

**Comment by nikhiljindal** [comment]
Yes including multiple versions sgtm.
We also need it for the e2e framework to be able to use the generated clientset.

Why include only two stable versions and why not include all of them? I can see the same problem coming again with some controller wanting to use a new version not included in the clientset, if we restrict it to only 2.

**Comment by caesarxuchao** [comment]
I think we will need a deprecation policy. We can discuss this later though, now only `batch` has v1 and v2alpha1.

**Comment by soltysh** [comment]
@caesarxuchao omg, I just hit that problem, again. Your solution is like a salvation, and has my full support :+1: 

Re the change itself, why instead of having `BatchV1`, `BatchV2Alpha`, etc. not having `Batch` with an argument being the version? Passing nil give you a default.

**Comment by caesarxuchao** [comment]
It's to satisfy the type system, the returned interface will be different for different argument.

**Comment by soltysh** [comment]
Right... forgot about that one, thanks.

**Comment by deads2k** [comment]
I think that the unversioned client should be able to negotiate against known server version it has conversions for which allow it to return to the hub.  I don't think that the unversioned client should have to worry about versions at all and it should prioritize them in the order that discovery returns them.

I'm ok with having the non-converting clients have versioned flavors, but if we do that I think we should always be forced to specify the version we want so that the transition from one version to two versions isn't overly painful for clients.

**Comment by caesarxuchao** [comment]
I experimented the idea locally. I actually didn't make the internalclientset (unversioned client) to include multiple versions, because it's hard to express that in [client-gen](https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/client-gen/main.go#L37-L47), so @soltysh might be disappointed. 

> unversioned client should be able to negotiate against known server version it has conversions 

@deads2k what's the use case of negotiating versions? I think we'll have this discussion in #35159 eventually. I can see it's useful for kubectl, because it might need to talk to different cluster with different versions enabled. 
I don't think other components need version negotiation. Controllers will be upgraded with the API server, and HA rolling update won't be a problem as I described in https://github.com/kubernetes/kubernetes/issues/35159#issuecomment-255236489. Kubelet and scheduler should be deprecated after two minor versions.
